### PR TITLE
Implement secp256k1 recovery native function

### PR DESCRIPTION
**Pull request by hendrikhofstadt**
## Motivation

Verifying secp256k1 will help create compatibility with the EVM world. E.g. this will allow 0x style signed orders to be compatible, existing bridges to adapt quicker and GSN style services to reuse infrastructure.

libsecp256k1 is pure rust, widely adopted and well tested.

Open for a discussion on gas pricing for this.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/aptos-labs/aptos-core/blob/main/CONTRIBUTING.md#pull-requests)?

CLA pending

## Test Plan

Tests written in Move

---
This change is [Reviewable](https://reviewable.io/reviews/aptos-labs/aptos-core/1305)

### CODE REVIEW

**Review by davidiw — COMMENTED** [review]
This is an amazing contribution. I greatly appreciate the steps you went through to build this. It isn't super clear to me exactly how this would be used. Granted I may not be as deep into the context that you provided. Could you help work out a clearer and more concrete application model?

**Comment by hendrikhofstadt** [comment]
The immediate motivation is that this will be required for an integration of Wormhole (Interoperability protocol).
Wormhole uses a ECDSA multisig on the secp256k1 curve only.

ECDSA on secp256k1 is the main schema and curve for ETH based chains. The `ecrecover` precompile/native in ETH is used by most developers for signature verification (by recovering the signer key) and has thus made it into most non-EVM execution layers as well (e.g. Solana, NEAR). For most protocols this curve is the lowest common denominator and they are used to the `ecrecover` as primary operation.
This native is a more low-level implementation of the public key recovery. ecrecover additionally hashes (keccak/sha3 256) the key and truncates the result (thereby deriving the ethereum address). This can be achieved in a higher level function since afaik sha3 is already supported as builtin.

Adopting this will make it easier for existing protocols to expand into aptos, developers to have a smoother transition finding well known primitives and bridges to adopt aptos without requiring them to switch to eddsa.

Other prominent examples of protocols that heavily rely on this are 0x (ecdsa signed limit orders) and Chainlink (oracle prices are signed using ecdsa)

File: aptos-move/framework/aptos-framework/sources/Signature.move [hunk 1]
```diff
@@ -21,4 +21,43 @@ module AptosFramework::Signature {
...
+    /// Returns `(public_key, true)` if inputs are valid and `([], false)` if invalid.
+    native public fun secp256k1_recover(
+        message: vector<u8>,
+        recovery_id: u8,
+        signature: vector<u8>
+    ): (vector<u8>, bool);
+
+    #[test]
+    /// Test on a valid signature created using pk = x"0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef"
+    fun test_secp256k1() {
```
Comment (alinush) [inline aptos-move/framework/aptos-framework/sources/Signature.move:36]: Can we add a randomized test here: 

- pick a random key-pair, 
- pick and sign a random message, 
- call recover
- check it returns the PK.

Comment (alinush) [reply aptos-move/framework/aptos-framework/sources/Signature.move:36]: ...otherwise, it looks good.

**Review by alinush — CHANGES_REQUESTED** [review]
Can we add a randomized test case to `test_secp256k1` (see comments)?

**Pull request merged**